### ISSUE TITLE
Ambiguous method call when using lambda in randomize method

### ISSUE THREAD

**Issue by petromir** [benas/random-beans#245]
Hi,

Using the following configuration:

```java
EnhancedRandomBuilder.aNewEnhancedRandomBuilder()
	.randomize(BigDecimal.class, () -> {
		// Some randomization logic			
	})).build();
```

the compiler throws:

```java
java: reference to randomize is ambiguous
  both method <T,R>randomize(java.lang.Class<T>,io.github.benas.randombeans.api.Randomizer<R>) in io.github.benas.randombeans.EnhancedRandomBuilder and method <T,R>randomize(java.lang.Class<T>,java.util.function.Supplier<R>) in io.github.benas.randombeans.EnhancedRandomBuilder match
```

**Comment by benas** [comment]
Hi,

Yes, it is ambiguous. The compiler cannot statically infer the type of your lambda. So you have to specify if it is a Supplier or a Randomizer:

```java
EnhancedRandom enhancedRandom = EnhancedRandomBuilder.aNewEnhancedRandomBuilder()
       .randomize(String.class, (Supplier<String>) () -> "foo")
       .randomize(Integer.class, (Randomizer<Integer>) () -> 10)
       .build();
```

Do you think this is an issue?

Kr
Mahmoud

**Comment by petromir** [comment]
Explicit type declaration makes the code more verbose (and as we know less is more 😄 ), taking into account that `Randomizer` and `Supplier` have the same definition, with just different method name. In my opionion `Randomizer` is redundant in that case.

**Comment by benas** [comment]
Sure! But type inference and overloading do not cope together for lambdas. So a choice has to be made:

* You suggest to sacrifice overloading in favour of type inference
* We decided to sacrifice type inference in favour of overloading

When the first request to add support for suppliers came in, we decided to keep both randomizers and suppliers for two reasons:

* We didn't want people using random beans v2.x to rewrite their randomizers and use suppliers
* More flexibility and choice

But as always, it's a tradeoff, and we preferred backward compatibility and flexibility over type inference for lambdas (note this is only for lambdas).
